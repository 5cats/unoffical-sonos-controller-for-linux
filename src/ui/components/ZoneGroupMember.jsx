import React from 'react';

export function ZoneGroupMember(props) {
    return (
        <li>
            <span>{props.member.name}</span>
        </li>
    );
}

export default ZoneGroupMember;
