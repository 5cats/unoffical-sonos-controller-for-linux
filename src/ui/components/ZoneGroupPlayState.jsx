import { h } from 'preact';

export function ZoneGroupPLayState(props) {
    const { playState } = props;

    if (playState && playState.trackInfo && playState.trackInfo.title) {
        const icon = playState.isPlaying
            ? <i className="material-icons">play_arrow</i>
            : <i className="material-icons">pause</i>;
        let info = playState.trackInfo.title;

        if (playState.trackInfo.artist) {
            info = info + ' - ' + playState.trackInfo.artist;
        }

        return (
            <div className="play-state">
                {icon} {info}
            </div>
        );
    }

    return <div className="play-state" />;
}

export default ZoneGroupPLayState;
