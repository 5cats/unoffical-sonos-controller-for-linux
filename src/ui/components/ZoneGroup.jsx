import register from '../helpers/registerComponent';
import model from '../model';
import ZoneGroupMember from './ZoneGroupMember'; 

class ZoneGroup {

	componentDidMount () {
		var self = this;

		model.observe('currentZone', function () {
			self.forceUpdate();
		});
	}

	componentWillUnmount () {
		// unsubscribe
	}

	render () {

		var zoneNodes = this.props.data.ZoneGroupMember.map(function (z) {
			return (
				<ZoneGroupMember data={z.$} />
			);
		});

		var classString = 'not-selected'

		if(model.currentZone.$.ID === this.props.data.$.ID) {
			classString = 'selected';
		}

			return (
				<ul className={classString} onClick={this._onClick}>
					{{zoneNodes}}
				</ul>
		);
	}

	_onClick () {
		model.currentZone = this.props.data;
	}
}

ZoneGroup.prototype.displayName = "ZoneGroup";
export default register(ZoneGroup);