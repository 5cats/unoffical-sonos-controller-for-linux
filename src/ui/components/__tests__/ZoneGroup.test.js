import { h } from 'preact';
import { ZoneGroup } from '../ZoneGroup';
import ZoneGroupPlayState from '../ZoneGroupPlayState';
import ZoneGroupMember from '../ZoneGroupMember';
import { deep } from 'preact-render-spy';

jest.mock('../ZoneGroupMember');
ZoneGroupMember.mockReturnValue(<p className="member" />);

jest.mock('../ZoneGroupPlayState');
ZoneGroupPlayState.mockReturnValue(<p className="play-state" />);

describe('ZoneGroup', () => {
    let props;

    beforeEach(() => {
        props = {
            group: [
                {
                    member: 1,
                    coordinator: 'true',
                    host: 'myhost'
                }
            ],
            selectGroup: jest.fn(),
            showManagement: jest.fn(),
            playStates: {
                myhost: {
                    playing: true
                }
            }
        };
    });

    it('matches snapshot', () => {
        const context = deep(<ZoneGroup {...props} />);
        expect(context.output()).toMatchSnapshot();
        expect(context.find('.not-selected').length).toBe(1);
        expect(ZoneGroupPlayState).toHaveBeenCalled();

        const [lastCall] = ZoneGroupPlayState.mock.calls;
        expect(lastCall[0]).toMatchObject({
            playState: {
                playing: true
            }
        });
    });

    it('matches snapshot when selected', () => {
        const currentHost = 'myhost';

        props = {
            ...props,
            currentHost
        };

        const context = deep(<ZoneGroup {...props} />);
        expect(context.output()).toMatchSnapshot();
        expect(context.find('.selected').length).toBe(1);
    });

    it('clicking whole group selects it', () => {
        const context = deep(<ZoneGroup {...props} />);
        context.find('.zone-group').simulate('click');
        expect(props.selectGroup).toHaveBeenCalled();
    });

    it('clicking button opens management', () => {
        const context = deep(<ZoneGroup {...props} />);

        context.find('.group-button').simulate('click', {
            preventDefault: jest.fn(),
            stopPropagation: jest.fn()
        });

        expect(props.showManagement).toHaveBeenCalled();
    });
});
