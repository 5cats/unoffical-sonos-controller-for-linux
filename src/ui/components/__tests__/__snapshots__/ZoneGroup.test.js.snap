// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ZoneGroup matches snapshot 1`] = `
<ZoneGroup
  group={
    Array [
      Object {
        "coordinator": "true",
        "host": "myhost",
        "member": 1,
      },
    ]
  }
  playStates={
    Object {
      "myhost": Object {
        "playing": true,
      },
    }
  }
  selectGroup={[MockFunction]}
  showManagement={[MockFunction]}
>
  <div
    className="not-selected zone-group"
    onClick={[Function]}
  >
    <ul>
      <Component
        key="0"
        member={
          Object {
            "coordinator": "true",
            "host": "myhost",
            "member": 1,
          }
        }
      >
        <p
          className="member"
        />
      </Component>
    </ul>
    <div
      className="group-button"
      onClick={[Function]}
    >
      Group
    </div>
    <ZoneGroupPLayState
      playState={
        Object {
          "playing": true,
        }
      }
    >
      <p
        className="play-state"
      />
    </ZoneGroupPLayState>
  </div>
</ZoneGroup>
`;

exports[`ZoneGroup matches snapshot when selected 1`] = `
<ZoneGroup
  currentHost="myhost"
  group={
    Array [
      Object {
        "coordinator": "true",
        "host": "myhost",
        "member": 1,
      },
    ]
  }
  playStates={
    Object {
      "myhost": Object {
        "playing": true,
      },
    }
  }
  selectGroup={[MockFunction]}
  showManagement={[MockFunction]}
>
  <div
    className="selected zone-group"
    onClick={[Function]}
  >
    <ul>
      <Component
        key="0"
        member={
          Object {
            "coordinator": "true",
            "host": "myhost",
            "member": 1,
          }
        }
      >
        <p
          className="member"
        />
      </Component>
    </ul>
    <div
      className="group-button"
      onClick={[Function]}
    >
      Group
    </div>
    <ZoneGroupPLayState
      playState={
        Object {
          "playing": true,
        }
      }
    >
      <p
        className="play-state"
      />
    </ZoneGroupPLayState>
  </div>
</ZoneGroup>
`;
